src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -726,7 +726,36 @@ internal override void Complete()
 
                     this.Command.Context.SetVariable(SpecialVariables.LastExitCodeVarPath, _nativeProcess.ExitCode);
                     if (_nativeProcess.ExitCode != 0)
+                    {
                         this.commandRuntime.PipelineProcessor.ExecutionFailed = true;
+
+                        bool nativeCommandUseErrorActionPref = 
+                            this.Command.Context.GetBooleanPreference(
+                                SpecialVariables.NativeCommandUseErrorActionPreferenceVarPath, 
+                                false, 
+                                out _);
+
+                        // If $PSNativeCommandUseErrorActionPreference is $true, then evaluate $ErrorActionPreference
+                        if (nativeCommandUseErrorActionPref)
+                        {
+                            ActionPreference errorActionPref =
+                                this.Command.Context.GetEnumPreference(
+                                    SpecialVariables.ErrorActionPreferenceVarPath,
+                                    ActionPreference.SilentlyContinue,
+                                    out _);
+
+                            if (errorActionPref != ActionPreference.Ignore)
+                            {
+                                var errorRecord = 
+                                    new ErrorRecord(
+                                        exception: new Exception($"Program {this.NativeCommandName} ended with non-zero exit code {_nativeProcess.ExitCode}"), 
+                                        errorId: "NativeCommandThrow", 
+                                        errorCategory: ErrorCategory.NotSpecified,
+                                        targetObject: this.NativeCommandName);
+                                this.commandRuntime._WriteErrorSkipAllowCheck(errorRecord, isNativeError: true);
```

Comment:
Since we have isNativeError flag why not move the new code to _WriteErrorSkipAllowCheck()?

Reply:
I'll look into that.

Reply:
Yeah this was kind of a complicated question to answer, but I think it's worth us taking a look